### PR TITLE
gix-ref: Add fuzz-harness's

### PR DESCRIPTION
**Pull request by silvergasp**

### CODE REVIEW

File: gix-ref/fuzz/fuzz_targets/fuzz_packed_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+use libfuzzer_sys::fuzz_target;
+use std::fs;
+use std::hint::black_box;
+use std::path::PathBuf;
+
+struct FuzzedBufferPath(PathBuf);
+
+impl FuzzedBufferPath {
+    fn new() -> Self {
+        Self(PathBuf::from(format!("/tmp/fuzz_packed_buffer_{}", std::process::id())))
```
Comment (Byron) [inline gix-ref/fuzz/fuzz_targets/fuzz_packed_buffer.rs:16]: I think this really must be `tempfile::NamedTempFile` or else it will be racy (or might become racy) depending on the way the fuzzer parallelises. Right now it assumes process ids are enough and I don't know if that's always going to be true.

For safety, using clash-free tempfile seems like the better choice.

**Review by Byron — COMMENTED** [review]
Thanks a lot!

There is just a question about the safety of relying on the process ID alone for filenames.

Further, seeing the workspace adjustment, there is a [related question](https://github.com/Byron/gitoxide/issues/703#issuecomment-1859643133) that fits in here quite well.

Thanks for your help!

File: gix-ref/fuzz/fuzz_targets/fuzz_packed_buffer.rs (hunk 1 above)
Comment (Byron) [reply gix-ref/fuzz/fuzz_targets/fuzz_packed_buffer.rs:16]: > How do you feel about me refactoring [open](https://github.com/Byron/gitoxide/blob/0fe20e80145419e1662f869657dabf689786395f/gix-ref/src/store/packed/buffer.rs#L33) into two functions one that opens from a `&[u8]` and is memory backed and the other which maintains the same api i.e. dynamically chooses memory Mapped/InMemory backing.

A great idea! This is the way to go :).

Comment (silvergasp) [reply gix-ref/fuzz/fuzz_targets/fuzz_packed_buffer.rs]: Done.

File: gix-ref/src/store/packed/buffer.rs [hunk 2]
```diff
@@ -84,6 +68,35 @@ pub mod open {
...
+            };
+            Self::open_with_backing(backing, path)
+        }
+
+        /// Open a buffer from existing memory.
+        ///
+        /// In order to allow fast lookups and optimizations, the contents of the packed refs must be sorted.
+        /// If that's not the case, they will be sorted on the fly with the data being written into a memory buffer.
+        #[cfg(fuzzing)]
+        pub fn open_from_memory(bytes: &[u8]) -> Result<Self, Error> {
```
Comment (Byron) [inline gix-ref/src/store/packed/buffer.rs:96]: As this adds a feature, could you put this change into a separate commit and prefix the subject line with `feat:` ? The subject itself could say which method was added, with the body being a short description of what it does. 

That automatically goes into the changelog.

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks a lot! 

As I am asking for a commit to be inserted, could you also avoid having `gix-ref:` prefixes? They should be removed without replacement, as it would cause trouble with the changelog.

Thanks again.

File: gix-ref/src/store/packed/buffer.rs (hunk 2 above)
Comment (silvergasp) [reply gix-ref/src/store/packed/buffer.rs:96]: Done

**Comment by silvergasp** [comment]
> They should be removed without replacement, as it would cause trouble with the changelog.

Oops I misread this. And replaced the prefixes with 'test'. If that's going to cause problems I'll completely remove the prefixes.

**Review by Byron — APPROVED** [review]
Thanks a lot!

I have made a few adjustments myself, with one of them being https://github.com/Byron/gitoxide/pull/1188/commits/5828f3796031de8d9f50bd6536697754da0b3d8c that you might want to have a look at in case there a good reason to keep using Cargo.lock files for fuzzers as well.

The problem I see with these lock files is that I will never update them, so they are bound to use outdated and possibly broken dependencies of crates that can also cause failures. Maybe I am missing something though.
